src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java

```diff
@@ -359,9 +363,23 @@ public ProgrammingExercise setupProgrammingExercise(ProgrammingExercise programm
 
         versionControlService.get().addWebHook(testsRepoUrl, ARTEMIS_BASE_URL + TEST_CASE_CHANGED_API_PATH + programmingExercise.getId(), "Artemis Tests WebHook");
 
+        // Depending on the activated VCS/CI systems, the VCS system pushes commit notifications to the CI, or the CI pulls
+        addOptionalNotificationsIfNecessary(projectKey, templateParticipation, solutionParticipation, exerciseRepoUrl, testsRepoUrl, solutionRepoUrl);
+
         return programmingExercise;
     }
 
+    private void addOptionalNotificationsIfNecessary(String projectKey, TemplateProgrammingExerciseParticipation templateParticipation,
+            SolutionProgrammingExerciseParticipation solutionParticipation, URL exerciseRepoUrl, URL testsRepoUrl, URL solutionRepoUrl) {
+        final var templatePlanNotificationUrl = continuousIntegrationService.get().getWebhookUrl(projectKey, templateParticipation.getBuildPlanId());
+        final var solutionPlanNotificationUrl = continuousIntegrationService.get().getWebhookUrl(projectKey, solutionParticipation.getBuildPlanId());
+        if (templatePlanNotificationUrl.isPresent() && solutionPlanNotificationUrl.isPresent()) {
+            versionControlService.get().addWebHook(exerciseRepoUrl, templatePlanNotificationUrl.get(), "Artemis Exercise WebHook", CI_TOKEN);
```

Comment:
same comment as above. I don't think we should use `CI_TOKEN` here. Please move this kind of logic into the concrete VCS Service